Finalize filtering in `listTransactions` by converting `Range UTCTime` to `Range SlotId`

Anviking opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->
#466 

# Overview

Previous PR #588 added filtering-support to `DB.readTxHistory` using `Range SlotId`. Here we do the final connecting to allow `WalletLayer.listTransactions` filtering by `Range UTCTime`.

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [ ] I have added a `slotAtTime` mirroring `slotStartTime`. I've added an unfinished roundtrip test for this. ⚠️ I'm not sure this is exactly how we want it. There might be subtleties regarding to which slot the boundary belongs to etc.
- [ ] TODO: Convert `Range UTCTime` to `Range SlotId` argument `listTransactions` (Maybe `Range` could be functor or something?)


# Comments

- I've tried to separate into confined commits
- Doing the `UTCTime -> SlotId` conversion _might_ not be needed in the future, depending on how we deal with proper slotting and changing epoch/slot-lengths.

<!-- Additional comments or screenshots to attach if any -->

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
WIP: Add slotAtTime :: ... -> UTCTime -> SlotId

Anviking committed:
WIP: Use the range arg in listTransactions for filtering

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -146,6 +154,13 @@ spec = do
+            property $ \slotLength epochLength startTime sl -> do
+                let slotAtTime' = slotAtTime epochLength slotLength startTime
+                let slotStartTime' = slotStartTime epochLength slotLength startTime
+                slotAtTime' (slotStartTime' sl) === sl
```
Currently getting a fascinating failure with an epoch length of 1, making `0` the only slotNumber.

I guess we could generate flat slots / "SlotNo"s and convert them.
```
  test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:158:9:
  1) Cardano.Wallet.Primitive.Types, SlotId <-> UTCTime conversions, slotAtTime . slotStartTime == id
       Falsifiable (after 1 test and 2 shrinks):
         SlotLength 1s
         EpochLength 1
         StartTime 1858-11-17 00:00:00 UTC
         SlotId {epochNumber = 3, slotNumber = 33}
         SlotId {epochNumber = 36, slotNumber = 0} /= SlotId {epochNumber = 3, slotNumber = 33}
```

Anviking commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -499,3 +514,22 @@ instance Arbitrary WalletName where
+
+deriving via UTCTime instance Arbitrary StartTime
+deriving via (NonZero NominalDiffTime) instance Arbitrary SlotLength
+deriving via (NonZero Word64) instance Arbitrary EpochLength
```
Couldn't resist trying 😅 

We need https://www.stackage.org/lts-13.30/package/stylish-haskell-0.9.2.2 to get CI to accept this, which I think maybe is straight-forward to get if we just bump LTS?

Anviking replied:
Also, the shrinking might be wrong / non-terminating 🤔

KtorZ replied:
> maybe is straight-forward to get if we just bump LTS?

It's never straightforward to bump the LTS :sweat_smile: ... Always bring tons of problems with the CI and caches that get us busy for half a day usually. But, sure, we'll have to eventually.

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -759,10 +759,11 @@ newWalletLayer tracer bp db nw tl = do
                 ErrListTransactionsStartTimeLaterThanEndTime $
                     ErrStartTimeLaterThanEndTime start end
             _ -> pure ()
+        let slotIdRange = (error "TODO :: UTCTime -> SlotId") timeRange
```
Re 
```haskell
case (mStart, mEnd) of
            (Just start, Just end) -> when (start > end) $ throwE $
	                ErrListTransactionsStartTimeLaterThanEndTime $
	                    ErrStartTimeLaterThanEndTime start end
	            _ -> pure ()
```

I considered moving this logic to `Primitive.Types` as `isValid :: Range a -> Bool`, but then there is no nice way of constructing `ErrStartTimeLaterThanEndTime` with pure `UTCTime`s.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/Primitive/TypesSpec.hs:
```diff
@@ -499,3 +514,22 @@ instance Arbitrary WalletName where
+        . filter (/= 0)
+        $ shrink x
+      where
+        suggestOne shrinks = if shrinks == [] then [1] else []
```
Ever heard of QuickCheck's modifiers :grimacing: ?

-> https://hackage.haskell.org/package/QuickCheck-2.12/docs/Test-QuickCheck.html#t:NonZero

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -876,6 +877,13 @@ slotStartTime epochLength (SlotLength slotLength) (StartTime start) sl =
+slotAtTime epochLength (SlotLength slotLength) (StartTime start) time =
+    fromFlatSlot
+        epochLength
+        (round $ (time `diffUTCTime` start) / slotLength)
```
No sure about the `round` here :thinking:  ... Shouldn't it be `floor` ?  Like, if we have let's say a slotlength or 20s and, 32 seconds, we should be in slot 1, not 2. Though, this uses `fromFlatSlot` which also does another round of conversions. 

What about computing the slot and the epoch directly from the time ?

```hs
t = time `diffUTCTime` start
epochDuration = epochLength * slotLength
epochNumber = floor (t / epochDuration)
slotNumber = floor ((t - epochNumber * epochDuration) / slotLength) 
```

I think... ?

KtorZ replied:
Note that I am pretty sure we already do something similar in the restoration benchmark code.

KtorZ committed:
review 'slotAt' function with computation that's easier to verify

KtorZ committed:
represent 'EpochLength' with Word16 instead of Word64

We represent slot number in 'SlotId' as 'Word16', so it doesn't make much sense to allow the 'EpochLength' to go
far beyond that bound. This might bite us later in rounding and overflowing. So, better keep the epoch length the same
size as the slot number

KtorZ committed:
review arbitrary instances for SlotLength, EpochLength and StartTime

We cannot generate 'SlotId' and 'EpochLength' independently, as this may generate invalid data..

KtorZ committed:
review guard for range in list transaction

Make sure the guard is triggered before we actually go fetch the wallet in the database.

KtorZ committed:
implement slotId range for listTransaction using UTCTime range

KtorZ committed:
Re-generate nix machinery

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -376,8 +378,8 @@ data ErrListTransactions
-    { startTime :: UTCTime
-    , endTime :: UTCTime
+    { errStartTime :: UTCTime
+    , errEndTime :: UTCTime
```
renamed to avoid name clash.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -460,10 +462,10 @@ newWalletLayer tracer bp db nw tl = do
-        (SlotLength slotLength)
-        slotsPerEpoch
+        slotLength
+        epochLength
```
renamed for consistency in naming with the Primitive.Types module.

KtorZ commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -753,17 +756,26 @@ newWalletLayer tracer bp db nw tl = do
         -> SortOrder
         -> ExceptT ErrListTransactions IO [TransactionInfo]
     _listTransactions wid mStart mEnd order = do
+        guardRange (mStart, mEnd)
```
factored out for readability.

paweljakubas reviewed (approved):
LGTM
